### PR TITLE
[gTest] Fix environment variables not being set in CICD

### PR DESCRIPTION
**Pull request by BrianHarrisonAMD**
Environment variables were not being set correctly in the gTest CMake before running the test suite.
This was leading to tests running in pipeline stages that should ignore them, or skipping tests that should be ran.

For example:
- Full Tests - Fp32 Hip All gfx90A stage:
    - Before the change was skipping 1458 tests:
    image
    - After the fix is skipping 858 tests:
    image
- Smoke Fp32 - Fp32 Hip gfx90A stage:
    - Before the change was skipping 1458 tests (same as the all):
    image
    - After the change is skipping 1537 tests:
    image

Issue was due to the following cmake issues with gtest_discover_tests:
- https://gitlab.kitware.com/cmake/cmake/-/issues/23120
- https://gitlab.kitware.com/cmake/cmake/-/issues/17812

With the environment variables fixed, there was also several tests that were failing and needed to be fixed:
- The adam tests were failing due to warnings being treated as errors in the build.
    - Warning for float comparison with ==.
    - Warning for not specifying values used in Macros for float_types.h used in MIOpenAdam.cpp from #2868 
- The layernorm tests were failing due to warnings being treated as errors for unused variables used in MIOpenLayerNorm.cpp from #2833 .

### CODE REVIEW

**Comment by BrianHarrisonAMD** [comment]
Greetings @et16kr!

If you get a chance, can you review the changes in MIOpenAdam.cpp, and adam.cpp to ensure they match your intention? With the environment variable fixes, the adam tests started running in CICD, and there were a few warnings that were being treated as errors that needed to be resolved.

**Comment by BrianHarrisonAMD** [comment]
Greetings @seungmanhan!

If you get a chance, can you review the changes in MIOpenLayerNorm.cpp? With the environment variable fixes, the layernorm tests started running in CICD, and there was a simple warning that was being treated as an error that needed to be resolved.

**Review by seungmanhan — APPROVED** [review]
I confirmed the t5layernorm changes.

**Review by et16kr — APPROVED** [review]
I confirmed the adam changes.

**Review by CAHEK7 — APPROVED** [review]

File: test/gtest/CMakeLists.txt
Comment (atamazov) [inline test/gtest/CMakeLists.txt]: This is unreadable. Is is possible so split and reformat this line?

Comment (atamazov) [inline test/gtest/CMakeLists.txt]: `MIOPENDRIVER_MODE_CONV`, `MIOPENDRIVER_MODE_GEMM`, `MIOPENDRIVER_MODE_BN`, `MIOPENDRIVER_MODE_POOL` should not be exported, these are intended for use only in CMakeLists.txt (just like macros in C, to have some single place where these command line options of the driver are defined for use in this script).

gtests should have their own definitions. To be able to work standalone, at least.

**Review by atamazov — CHANGES_REQUESTED** [review]

File: test/gtest/CMakeLists.txt
Comment (BrianHarrisonAMD) [reply test/gtest/CMakeLists.txt]: Ah good point.

Comment (BrianHarrisonAMD) [reply test/gtest/CMakeLists.txt]: Updated, please check again when you get a chance!

Comment (BrianHarrisonAMD) [reply test/gtest/CMakeLists.txt]: Updated, please check again when you get a chance!

I removed the env variables for MIOPENDRIVER_MODE_X, and updated the tests to be split to enable running them standalone properly.

File: test/gtest/miopendriver_common.hpp
Comment (atamazov) [inline test/gtest/miopendriver_common.hpp]: [Recommendation][style]
```suggestion
#define MIOPENDRIVER_BASEARG_CONV_FLOAT "conv"
```
here and below. Or something like
```C++
namespace miopendriver {
namespace basearg {
namespace conv {
const char float[] = "conv";
...
}}}
```

**Review by atamazov — COMMENTED** [review]
[Notice] This PR resolves https://github.com/ROCm/MIOpen/pull/2984#discussion_r1624770196 and https://github.com/ROCm/MIOpen/pull/2984#discussion_r1622927012.

File: test/gtest/miopendriver_common.hpp
Comment (BrianHarrisonAMD) [reply test/gtest/miopendriver_common.hpp]: Agreed, and updated.

File: test/gtest/miopendriver_common.hpp
Comment (amberhassaan) [inline test/gtest/miopendriver_common.hpp]: please prefer C++ style constant definitions to `#define`.

File: test/gtest/miopendriver_conv2d_trans.cpp
Comment (amberhassaan) [inline test/gtest/miopendriver_conv2d_trans.cpp]: I don't quite understand this logic (as I haven't worked with this code). I'd appreciate a descriptive comment that explains what `standalone` mode means etc.

**Review by amberhassaan — COMMENTED** [review]
a couple of minor comments.

File: test/gtest/miopendriver_common.hpp
Comment (BrianHarrisonAMD) [reply test/gtest/miopendriver_common.hpp]: Does the updated version look good?

File: test/gtest/miopendriver_conv2d_trans.cpp
Comment (BrianHarrisonAMD) [reply test/gtest/miopendriver_conv2d_trans.cpp]: Updated comment to be more descriptive. 
Please take a look when you get a chance, and let me know if that helps clarify!

**Review by iq136boy — APPROVED** [review]

File: test/gtest/miopendriver_conv2d_trans.cpp
Comment (amberhassaan) [reply test/gtest/miopendriver_conv2d_trans.cpp]: ok, this makes sense. Thank you.

File: test/gtest/miopendriver_common.hpp
Comment (amberhassaan) [inline test/gtest/miopendriver_common.hpp:47]: This is fine for now. In reality we should be composing it with `{conv, pool, gemm} x {f32, f16, bf16, int8, ...}`, and there should be a separate commandline option for specifying the data type. Out of scope of this PR obviously.

File: test/gtest/miopendriver_regression_big_tensor.cpp
Comment (amberhassaan) [inline test/gtest/miopendriver_regression_big_tensor.cpp]: A logical follow up step would be to pick these and the `MIOPEN_DECLARE_ENV...` calls into a common header file and then you have to write the comment only once :).

Comment (BrianHarrisonAMD) [reply test/gtest/miopendriver_regression_big_tensor.cpp]: Agreed! I believe we made a separate task for doing that in bulk across the tests, but could take the first step now if it makes sense. Thoughts?

Comment (amberhassaan) [reply test/gtest/miopendriver_regression_big_tensor.cpp]: I say start now and then you or someone else can finish it up in a follow up PR. That's my personal preference.

Comment (BrianHarrisonAMD) [reply test/gtest/miopendriver_regression_big_tensor.cpp]: @amberhassaan Updated, can you check to see if you like the changes? 

Thanks!

Comment (amberhassaan) [reply test/gtest/miopendriver_regression_big_tensor.cpp]: Yes, this is great. Thank you.

**Review by amberhassaan — APPROVED** [review]

**Comment by junliume** [comment]
@BrianHarrisonAMD please resolve the conflicts, we want to push this PR forward asap. Thanks!

**Comment by BrianHarrisonAMD** [comment]
@junliume conflicts should be resolved now, and it should be ready to merge.
I was waiting on a full CI run before adding the tag, but I believe the previous run from before the merge is almost finished.

**Comment by BrianHarrisonAMD** [comment]
@junliume Should be good to merge now! CI passing, and up to date. 

Thanks!

**Review by junliume — APPROVED** [review]

**Pull request merged**
